src/XrdTls/XrdTlsContext.cc

```diff
@@ -860,32 +860,30 @@ void *XrdTlsContext::Session()
 //
    DBG_CTX("Replacing x509 store with new contents.");
 
+   //The new content has been generated by the Refresh thread. So we must use it
+   XrdTlsContext *xrdctxnew = pImpl->ctxnew;
+
 // Get the new store and set it in our context. Setting the store is black
 // magic. For OpenSSL < 1.1, Two stores need to be set with the "set1" variant.
 // Newer version only require SSL_CTX_set1_cert_store() to be used.
 //
-   X509_STORE *newX509 = SSL_CTX_get_cert_store(pImpl->ctx);
+   X509_STORE *newX509 = SSL_CTX_get_cert_store(xrdctxnew->pImpl->ctx);
 
 #if OPENSSL_VERSION_NUMBER < 0x10101000L
-   SSL_CTX_set1_verify_cert_store(pImpl->ctx, newX509);
-   SSL_CTX_set1_chain_cert_store(pImpl->ctx, newX509);
+   SSL_CTX_set1_verify_cert_store(xrdctxnew->pImpl->ctx, newX509);
+   SSL_CTX_set1_chain_cert_store(xrdctxnew->pImpl->ctx, newX509);
 #else
-   SSL_CTX_set1_cert_store(pImpl->ctx, newX509);
+   SSL_CTX_set1_cert_store(xrdctxnew->pImpl->ctx, newX509);
 #endif
 
-// Save the generated context and clear it's presence
-//
-   XrdTlsContext *ctxold = pImpl->ctxnew;
-   pImpl->ctxnew = 0;
-
```

Comment:
OK, I see there is a misunderstanding here. What the original code actually does is grab the X509 store from the new context and sets it to be used by the current context. It then simply deletes the new context and sets the pointer to it to zero, indicating that there is no new x509 store so that the x509 transfer executes only once and allows a new x509 store to be created. The reason we need to do that is because an x509 store creation also requires the creation of a context but we never want to use that newly created context. We must continue to use the current context because we don't know how the current context is being used (e.g. there may be an outstanding pointer to it). So, we must always use the current context to create SSL objects. The changes starting on line 876  undo all of that and I would say that it's luck that it worked. Hence, all of the original code starting at line 876 should not have been changed. The code above that line was indeed incorrect and your changes fixed that, thanks!.